### ISSUE TITLE
Feature request: dependency between serverless services

### ISSUE THREAD

**Issue by danielesegato** [serverless/serverless#1618]
##### Feature Request:

Ability to publish a serverless project and to add it as dependency.
It should use the same profile and stage of the current project.
It has to be deployed before the current serverless project.
##### Benefits:

To explain why I ask this I have to explain what I did.

> In our project we needed to deploy a CloudSearch Domain withing the cloud formation resources.
> 
> As you know CloudSearch is not supported by CloudFormation.
> So I created a custom resource to handle it.
> 
> Problems:
> - the custom lambda was too big / complex to fit in the CloudFormation script
> - the domain takes around 10 minutes after it has been created to provide index and search URL
> 
> For the first issue I could put a zip file of the lambda somewhere but then I had a manual step to perform before I could be able to deploy my project.
> 
> For the second a lambda was no good, I put in place a polling mechanism with CloudWatchEvents and a dedicated lambda to check for the domain and complete the CloudFormation resource creation.
> 
> To do all this I created a side-serverless project that deploy all the resources needed for handling the creation of CloudSearch domains. This project contained the lambda to create, update and delete the domain (completely configurable via CloudFormation parameters).
> 
> Now to deploy my project I just need to deploy this side serverless project on the same AWS account and region and deploy the lambda.
> 
> From the main project I just need the lambda name to setup my custom CloudSearch resource and then I can just create my project with domain and everything I need.
> The first deploy takes 10 minutes but that can't be helped.
> 
> This is still a manual step to perform but at least I've all in the same repository and it only requires executing a simple script.

If I could package my side-serverless-project and **reference it as dependency** I could make this seamless and also **reuse my side serverless project** (which is completely generic for cloudsearch domain creation through cloud formation).
##### Drawbacks:

It surely add complexity to the project.

May not be the best solution to solve my problem. That's the best I could find.
##### Additional Details:

I also tough on writing a serverless plugin that created the domain at initialization.
This way the pc launching the script could wait as much time as he wanted for the domain to be created (no need for polling).

But this involve creating AWS resources before even deploying the cloud formation template. Also requires injecting in the cloudformation template a custom resource lambda to fake the creation of the domain.

I had no other ideas :)
##### TLDR:

I build a side-serverless project to provide a lambda as service token for create a CloudSearch domain through a Cloud Formation custom resource.
Couldn't do this in the main project because of how it works.

The side-project is reusable and need to be deployed before the main project.

I wish I could just add it as dependency to the main project.
##### Discussion:

Completely open to receive insults or to be told I did something stupid and the solution was far easier then that :)

**Comment by flomotlik** [comment]
@danielesegato We actually have been thinking about this as well in the past and we're simply not sure yet how dependencies between services from V1 going forward will be handled.

I'll mark this one as discussion so we can capture ideas here going forward.

**Comment by thenikso** [comment]
Ok here is another (hypothetical) senario using API gateway:

I have various teams creating microservices. We decided that private resources will have a `/user/<name>` prefix so:

- one team will develop the foo project and use URLs like `/user/<name>/foo/<id>`
- another team will develop the bar with URLs like `/user/<name>/bar/<id>`

Those project will eventually have a different API Gateway host which should then be problematic to handle with a custom domain.

The best way to handle this that I can think of right now would be to use Serverless only to deploy lambdas and create a single API Gateway api in another way.

For this particular case I'd expect to be able to specify the API Gateway name something like this:

```yaml
events:
  html:
    api: my-shard-api-name
    method: get
    path: /user/{userId}/foo/{fooId}
```

**Comment by flomotlik** [comment]
We're currently discussing more ways to do things in #2481 so I'm closing this issue here.
